opencensus/exporters/trace/ocagent/internal/ocagent_exporter.cc

```diff
@@ -0,0 +1,338 @@
+    // 2. span_id
+    to_span->set_span_id(from_span.context().span_id().ToHex());
+    // 3. parent_span_id
+    to_span->set_parent_span_id(from_span.parent_span_id().ToHex());
+    // 4. name
+    SetTruncatableString(from_span.name(), kDisplayNameStringLen,
+                         to_span->mutable_name());
+    // 5. The start time of the span.
+    EncodeTimestampProto(from_span.start_time(), to_span->mutable_start_time());
+    // 6. The end time of the span.
+    EncodeTimestampProto(from_span.end_time(), to_span->mutable_end_time());
+
+    // 7. Export Attributes
+    ConvertAttributes(from_span, to_span);
+
+    // 8. stack_trace
+
+    // 9. Export Time Events.
+    ConvertTimeEvents(from_span, to_span);
+
+    // 10. Export Links.
+    ConvertLinks(from_span, to_span);
+
+    // 11. The status of the span.
+    to_span->mutable_status()->set_code(
+        static_cast<int32_t>(from_span.status().CanonicalCode()));
+    to_span->mutable_status()->set_message(from_span.status().error_message());
+
+    // 12. True if the parent is on a different process.
+    to_span->mutable_same_process_as_parent_span()->set_value(
+        !from_span.has_remote_parent());
+
+    // 13. child_span_count
+
+    // 14. span kind
+
+    // 15. tracestate
+
+    // 16. resource
+    // (TODO) Add FULL support.
+  }
+}
+
+Handler::Handler(OcAgentOptions &&opts) : opts_(std::move(opts)) {
+  InitNode();
+  ConnectAgent();
+}
+
+void Handler::Export(
+    const std::vector<::opencensus::trace::exporter::SpanData> &spans) {
+  ::opencensus::proto::agent::trace::v1::ExportTraceServiceRequest request;
+  ConvertSpans(spans, &request);
+  request.mutable_node()->CopyFrom(nodeInfo_);
+
+  grpc::ClientContext context;
+  context.set_deadline(absl::ToChronoTime(absl::Now() + opts_.rpc_deadline));
+
+  auto stream = opts_.trace_service_stub->Export(&context);
+  bool ok = stream->Write(request);
+}
+
+void Handler::InitNode() {
+  nodeInfo_.New();
+  auto identifier = nodeInfo_.mutable_identifier();
+
+  char host[100] = {0};
+  if (gethostname(host, sizeof(host)) < 0) {
+    snprintf(host, sizeof(host), "____default_host_name");
+  }
+  identifier->set_host_name(host);
+  identifier->set_pid(getpid());
+
+  {
+    auto ts_ptr = identifier->mutable_start_timestamp();
+    auto t = absl::Now();
+    const int64_t s = absl::ToUnixSeconds(t);
+    ts_ptr->set_seconds(s);
+    ts_ptr->set_nanos((t - absl::FromUnixSeconds(s)) / absl::Nanoseconds(1));
+  }
+
+  auto library_info = nodeInfo_.mutable_library_info();
+  library_info->set_language(
+      ::opencensus::proto::agent::common::v1::LibraryInfo_Language_CPP);
+  library_info->set_exporter_version("0.0.1");
+  library_info->set_core_library_version("0.0.1");
+
+  auto service_info = nodeInfo_.mutable_service_info();
+  service_info->set_name(host);
+}
+
+void Handler::ConnectAgent() {
+  ::opencensus::proto::agent::trace::v1::ExportTraceServiceRequest request;
+  request.mutable_node()->CopyFrom(nodeInfo_);
```

Comment:
```*request.mutable_node() = nodeInfo_;```

Reply:
What's the difference between these ?

Reply:
> Assignment is preferred because CopyFrom isn't type safe.
> 
> e.g.:
> 
> ```
> request.mutable_node()->CopyFrom(nodeInfo_.library_info()); // compiles
> *request.mutable_node() = nodeInfo_.library_info(); // compile-time error
> 
> error: no match for 'operator=' (operand types are 'opencensus::proto::agent::common::v1::Node' and 'const opencensus::proto::agent::common::v1::LibraryInfo')
> ```

Got it! 👍